File: backend/src/main/java/com/smartcampus/backend/domain/Order.java
package com.smartcampus.backend.domain;

import jakarta.persistence.*;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import java.time.LocalDateTime;

@JsonIgnoreProperties({ "hibernateLazyInitializer", "handler" })
@Entity
@Table(name = "orders")
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "student_id", nullable = true)
    private User student;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "outlet_id", nullable = false)
    private Outlet outlet;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "pickup_slot_id", nullable = true)
    private PickupSlot pickupSlot;

    @Column(nullable = false, length = 30)
    private String status;
    // PLACED, PREPARING, READY, PICKED, EXPIRED, CANCELLED

    @Column(nullable = false)
    private double totalAmount;

    @Column(nullable = false, length = 20)
    private String paymentMode;
    // ONLINE, CASH

    @Column(nullable = false, length = 20)
    private String paymentStatus;
    // PENDING, PAID, FAILED

    // Smart predicted completion time
    @Column(nullable = false)
    private LocalDateTime readyAt;

    // Grace expiry time (readyAt + 30 mins)
    @Column(nullable = false)
    private LocalDateTime expiresAt;

    @Column(nullable = false)
    private LocalDateTime createdAt;

    /**
     * 4-digit OTP generated at order creation (COD) or at payment verification
     * (ONLINE).
     * Student shows this to the manager at pickup.
     * Manager submits it via POST /api/orders/{id}/pickup to mark the order PICKED.
     * Stored as a String to preserve any leading zeros (e.g. "0472").
     */
    @Column(length = 4)
    private String pickupOtp;

    /**
     * Source of the order:
     * "PLATFORM" = placed by student via app (normal flow)
     * "COUNTER" = placed by manager at the counter for a walk-in customer
     */
    @Column(name = "order_source", nullable = false, length = 10)
    private String orderSource = "PLATFORM";

    /**
     * For COUNTER orders: the walk-in customer's name typed by the manager.
     * Null for PLATFORM orders (student name is in the student relation).
     */
    @Column(length = 120)
    private String customerName;

    protected Order() {
    }

    public Order(User student,
            Outlet outlet,
            PickupSlot pickupSlot,
            String status,
            double totalAmount,
            String paymentMode,
            String paymentStatus,
            LocalDateTime readyAt,
            LocalDateTime expiresAt) {

        this.student = student;
        this.outlet = outlet;
        this.pickupSlot = pickupSlot;
        this.status = status;
        this.totalAmount = totalAmount;
        this.paymentMode = paymentMode;
        this.paymentStatus = paymentStatus;
        this.readyAt = readyAt;
        this.expiresAt = expiresAt;
        this.createdAt = LocalDateTime.now();
    }

    // ── Getters ───────────────────────────────────────────────────────────────
    public Long getId() {
        return id;
    }

    public User getStudent() {
        return student;
    }

    public Outlet getOutlet() {
        return outlet;
    }

    public PickupSlot getPickupSlot() {
        return pickupSlot;
    }

    public String getStatus() {
        return status;
    }

    public double getTotalAmount() {
        return totalAmount;
    }

    public String getPaymentMode() {
        return paymentMode;
    }

    public String getPaymentStatus() {
        return paymentStatus;
    }

    public LocalDateTime getReadyAt() {
        return readyAt;
    }

    public LocalDateTime getExpiresAt() {
        return expiresAt;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public String getPickupOtp() {
        return pickupOtp;
    }

    public String getOrderSource() {
        return orderSource;
    }

    public String getCustomerName() {
        return customerName;
    }

    // ── Setters ───────────────────────────────────────────────────────────────
    public void setPickupSlot(PickupSlot slot) {
        this.pickupSlot = slot;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public void setPaymentStatus(String ps) {
        this.paymentStatus = ps;
    }

    public void setPickupOtp(String pickupOtp) {
        this.pickupOtp = pickupOtp;
    }

    public void setOrderSource(String orderSource) {
        this.orderSource = orderSource;
    }

    public void setCustomerName(String n) {
        this.customerName = n;
    }
}

File: backend-fastapi/app/models/order.py
"""SQLAlchemy model for the `orders` table. Source: Order.java"""
from datetime import datetime
from typing import TYPE_CHECKING, Optional

from sqlalchemy import BigInteger, DateTime, Double, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base

if TYPE_CHECKING:
    from app.models.user import User
    from app.models.outlet import Outlet
    from app.models.pickup_slot import PickupSlot
    from app.models.order_item import OrderItem
    from app.models.payment import Payment
    from app.models.outlet_rating import OutletRating


class Order(Base):
    __tablename__ = "orders"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)

    # Nullable for COUNTER orders (walk-in, no student account)
    student_id: Mapped[Optional[int]] = mapped_column(
        BigInteger, ForeignKey("users.id"), nullable=True
    )
    outlet_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("outlets.id"), nullable=False
    )
    # Nullable for COUNTER orders
    pickup_slot_id: Mapped[Optional[int]] = mapped_column(
        BigInteger, ForeignKey("pickup_slots.id"), nullable=True
    )

    # PLACED | PREPARING | READY | PICKED | EXPIRED | CANCELLED
    status: Mapped[str] = mapped_column(String(30), nullable=False)
    total_amount: Mapped[float] = mapped_column(Double, nullable=False)
    # ONLINE | CASH
    payment_mode: Mapped[str] = mapped_column(String(20), nullable=False)
    # PENDING | PAID | FAILED
    payment_status: Mapped[str] = mapped_column(String(20), nullable=False)
    ready_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    expires_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)

    # 4-digit OTP stored as string to preserve leading zeros (e.g. "0472")
    pickup_otp: Mapped[Optional[str]] = mapped_column(String(4), nullable=True)

    # PLATFORM (student via app) or COUNTER (manager walk-in)
    # explicit @Column(name = "order_source") in Java — preserved
    order_source: Mapped[str] = mapped_column(String(10), nullable=False, default="PLATFORM")
    # Walk-in customer name for COUNTER orders
    customer_name: Mapped[Optional[str]] = mapped_column(String(120), nullable=True)

    # Relationships
    student: Mapped[Optional["User"]] = relationship(
        "User", foreign_keys=[student_id], lazy="joined"
    )
    outlet: Mapped["Outlet"] = relationship("Outlet", back_populates="orders", lazy="joined")
    pickup_slot: Mapped[Optional["PickupSlot"]] = relationship(
        "PickupSlot", back_populates="orders", lazy="select"
    )
    items: Mapped[list["OrderItem"]] = relationship("OrderItem", back_populates="order")
    payments: Mapped[list["Payment"]] = relationship("Payment", back_populates="order")
    rating: Mapped[Optional["OutletRating"]] = relationship(
        "OutletRating", back_populates="order", uselist=False
    )
